File: apps/backend-python/src/utils/jwt.py
from __future__ import annotations
from datetime import datetime, timedelta
from jose import jwt, JWTError
from src.config.env import env
import secrets

ALGORITHM = "HS256"


def parse_expiry(expiry_str: str) -> int:
    """Parse '7d', '4w', '1h' style strings into seconds."""
    unit = expiry_str[-1]
    value = int(expiry_str[:-1])
    multipliers = {"s": 1, "m": 60, "h": 3600, "d": 86400, "w": 604800}
    return value * multipliers.get(unit, 86400)

def sign_access_token(payload: dict) -> str:
    """Generate stateless JWT for Access Token (Short-lived)"""
    expires_delta = timedelta(seconds=parse_expiry(env.JWT_EXPIRES_IN))
    data = payload.copy()
    data["exp"] = datetime.utcnow() + expires_delta
    return jwt.encode(data, env.JWT_SECRET, algorithm=ALGORITHM)

def verify_token(token: str) -> dict | None:
    """Verify standard JWT Access Token"""
    try:
        payload = jwt.decode(token, env.JWT_SECRET, algorithms=[ALGORITHM])
        return payload
    except JWTError:
        return None

def generate_refresh_token() -> str:
    """Generate stateful opaque string for Refresh Token (Long-lived)"""
    return secrets.token_urlsafe(64)

File: apps/backend/src/utils/jwt.ts
import jwt, { type SignOptions } from 'jsonwebtoken';
import { env } from '../config/env';

export interface JWTPayload {
  id: string;
  email: string;
  name: string;
  role: string;
  profileImageUrl?: string;
}

export function signToken(payload: JWTPayload): string {
  const options: SignOptions = { expiresIn: env.JWT_EXPIRES_IN as SignOptions['expiresIn'] };
  return jwt.sign(payload, env.JWT_SECRET, options);
}

export function verifyToken(token: string): JWTPayload | null {
  try {
    return jwt.verify(token, env.JWT_SECRET) as JWTPayload;
  } catch {
    return null;
  }
}
